lib/core/notifier.js

```diff
@@ -272,51 +269,42 @@ class NotifierController {
    * @param {Request} request
    * @return {Promise}
    */
-  notifyDocumentUpdate (request) {
-    let
-      cacheId,
-      matchedRooms,
-      updatedDocument;
-
-    const { index, collection, _id } = request.input.resource;
-
-    // @todo do not make another get request, use the _source returned by update request
-    return this.storageEngine.get(index, collection, _id)
-      .then(result => {
-        updatedDocument = result;
-        matchedRooms = this._test(request, result._source, result._id);
-        cacheId = getCachePrefix(request) + updatedDocument._id;
-
-        const updatedFields = Object.keys(request.input.body)
-          .filter(_updatedFields => _updatedFields !== '_kuzzle_info');
-
-        return this.notifyDocument(matchedRooms, request, 'in', 'update', {
-          _id: updatedDocument._id,
-          _source: updatedDocument._source,
-          _updatedFields: updatedFields
-        });
-      })
-      .then(() => this.cacheEngine.get(cacheId))
-      .then(cachedRooms => {
-        if (cachedRooms !== null) {
-          const stopListening = _.difference(
-            JSON.parse(cachedRooms), matchedRooms);
+  async notifyDocumentUpdate (request, updatedDocument) {
 
-          return this.notifyDocument(stopListening, request, 'out', 'update', {
-            _id: updatedDocument._id
-          });
-        }
+    const
+      matchedRooms = this._test(
+        request,
+        updatedDocument._source,
+        updatedDocument._id);
+    const cacheId = getCachePrefix(request) + updatedDocument._id;
 
-        return null;
-      })
-      .then(() => {
-        if (matchedRooms.length > 0) {
-          return this._setCacheWithTTL(cacheId, JSON.stringify(matchedRooms));
-        }
+    const updatedFields = Object.keys(request.input.body)
+      .filter(_updatedFields => _updatedFields !== '_kuzzle_info');
 
-        return this.cacheEngine.del(cacheId);
-      })
-      .catch(error => this.kuzzle.log.error(error));
+    try {
+      await this.notifyDocument(matchedRooms, request, 'in', 'update', {
+        _id: updatedDocument._id,
+        _source: updatedDocument._source,
+        _updatedFields: updatedFields
+      });
+      const cachedRooms = await this.cacheEngine.get(cacheId);
+      if (cachedRooms !== null) {
+        const stopListening = _.difference(
+          JSON.parse(cachedRooms), matchedRooms);
+
+        await this.notifyDocument(stopListening, request, 'out', 'update', {
+          _id: updatedDocument._id
+        });
+      }
+
+      if (matchedRooms.length > 0) {
```

Comment:
Nitpicking: 
IMHO it's more clearer to have all the method returns at the same place (inside or outside the try catch)